### PR TITLE
Microcoded EA calculation

### PR DESCRIPTION
**Pull request by larsbrinkhoff**
This is a state machine for EA calculation.  It's inspired by microcode.  There is an `upc` which is a microcode PC, and `ucycles` specifies the length of the microprogram.

The microengine operates on two primary "registers".  `ea_address` is the accumulator for the effective address being calculated.  `ea_data` is the final operand, and also holds intermediate values read from memory.

A little work is done up-front when the state machine is initialised.  After that, the addressing modes are implemented using six microinstructions which all takes one microcycle (two CPU cycles):
- `n` does nothing,
- `r` reads one word (or byte) from memory,
- `p` fetches one word from the program (this should integrate with the prefetch registers),
- `a` transfers `ea_data` to `ea_address`,
- `sa` like `a` with sign-extension, and
- `idx` calculates the "indirect with index" addressing mode.

Byte and long operands are supported by masking the final result, or running the program two more microcyles, respectively.

### CODE REVIEW

File: cpuinstr/ea.c [hunk 1]
```diff
@@ -1,713 +1,182 @@
...
-static int rmw = 0;
-static int ea_prefetch_before_write = 0;
-static int last_register_change = 0;
+static LONG ea_address;
+static LONG ea_data;
+static LONG ea_mask;
+static WORD (*ea_read)(LONG);
 
-static BYTE ea_read_000_b(struct cpu *cpu, int reg)
+static LONG sign_ext_8(BYTE x)
```
Comment (stefanberndtsson) [inline cpuinstr/ea.c]: Should these be broken out to macros? They should be pretty much just as useful in instruction microcode as in the ea bits.

File: cpuinstr/ea.c [hunk 2]
```diff
@@ -1,713 +1,182 @@
...
 {
-  cpu->d[reg] = (cpu->d[reg]&0xffffff00)|data;
+  LONG addr = cpu->pc;
+  cpu->pc += 2;
+  return bus_read_word(addr);
 }
 
-static void ea_write_000_w(struct cpu *cpu, int reg, WORD data)
+// Do nothing.
+static void n(void)
```
Comment (stefanberndtsson) [inline cpuinstr/ea.c]: I'm not too fond of these single-char function names.

**Comment by stefanberndtsson** [comment]
Should read low word and read high word be different microcodes (and the same for write)? Maybe that's just unnecessary.

File: cpuinstr/ea.c (hunk 1 above)
Comment (larsbrinkhoff) [inline cpuinstr/ea.c]: Yes, and elsewhere also maybe?  They could go in `common.h`.

File: cpuinstr/ea.c (hunk 2 above)
Comment (larsbrinkhoff) [inline cpuinstr/ea.c]: Not to worry!  They started out inspired by the notation in the Yacht table.  But I'm moving them to a `ucode.c` which is not specific to EA calculation.

**Comment by larsbrinkhoff** [comment]
> Should read low word and read high word be different microcodes

Don't know!  They might, or they might not.  If you read the current version carefully, you can see that the `r` microop works for both reading a byte, a word, or a long, depending on context.

File: cpuinstr/ea.c (hunk 1 above)
Comment (stefanberndtsson) [inline cpuinstr/ea.c]: I put a few (used with MOVEM) into `cpu.h` earlier.
```
#define EXTEND_BYTE(byte) (((byte)&0x80) ? ((byte)|0xff00) : (byte))
#define EXTEND_BY2L(byte) (((byte)&0x80) ? ((byte)|0xffffff00) : (byte))
#define EXTEND_WORD(word) (((word)&0x8000) ? ((word)|0xffff0000) : (word))
```

Technically yours clean up unsigned more than mine do. :)
I'm all in favour of using `common.h` though. It seems logical that any part of the code could use it, not just the CPU.

**Comment by larsbrinkhoff** [comment]
Added support for writing back the result in a typical read-modify-write instruction.  This works by using `ea_address` from the reading phase.

**Comment by larsbrinkhoff** [comment]
Maybe some microcode explanation is in order.

    static uop_t index_uops[] = { n, p, idx, r, n, r, n };

This is the microprogram for the indexed addressing mode.  Before the microcode `ea_address` is initialised to the value of the base address register.  The Yacht table specifies "n np nr" for this mode, so the microprogram begins with an `n`.  The order inside the bus cycles are reversed from the Yacht notation, so the next step is a `p` fetch.  This reads the extension word and stores it in `ea_data`.  Then comes `idx`, which uses the extension word to add an offset and an index register to `ea_address`.  Finally, there are one or two bus cycles to read from `ea_address` to `ea_data`.

**Comment by larsbrinkhoff** [comment]
This is verified by modifying `CLR` to use the new EA calculation and booting to the Desktop.  This may not use all addressing modes, however, so there may well be bugs lurking.

File: cpuinstr/ea.c [hunk 3]
```diff
@@ -1,713 +1,274 @@
...
-  bus_write_byte(addr, data);
-  if(cpu_full_stacked_exception_pending() && (((mode&0x38)>>3) == 3 || ((mode&0x38)>>3) == 4)) {
-    cpu->a[mode&0x7] -= last_register_change;
-  }
-}
+static uop_t write_reg_uops[] = { n, n, n };
+static uop_t write_mem_uops[] = { w, n, w, n };
 
-void ea_write_word(struct cpu *cpu, int mode, WORD data)
+void ea_begin_modify(struct cpu *cpu, WORD op, LONG data, int c1, int c2, int c3, int c4)
```
Comment (stefanberndtsson) [inline cpuinstr/ea.c]: What's c1-c4? Comment? Better names?

**Comment by larsbrinkhoff** [comment]
This implements most of #122.

File: cpuinstr/ea.c (hunk 3 above)
Comment (larsbrinkhoff) [inline cpuinstr/ea.c]: Added comment and slightly better names.

**Comment by larsbrinkhoff** [comment]
Fixed BIG Demo.  There was a bug in the absolute short addressing mode.

Phalanx F5 crashes in master too, so it's not a blocker for this PR.

**Pull request merged**